src/examples/zkapps/dex/dex.ts

```diff
@@ -19,250 +19,339 @@ import {
   Struct,
   State,
   state,
+  UInt32,
 } from 'snarkyjs';
 
-export { Dex, DexTokenHolder, TokenContract, keys, addresses, tokenIds };
+export { createDex, TokenContract, keys, addresses, tokenIds };
 
 class UInt64x2 extends Struct([UInt64, UInt64]) {}
 
-class Dex extends SmartContract {
-  // addresses of token contracts are constants
-  tokenX = addresses.tokenX;
-  tokenY = addresses.tokenY;
+function createDex({
+  lockedLiquiditySlots,
+}: { lockedLiquiditySlots?: number } = {}) {
+  class Dex extends SmartContract {
+    // addresses of token contracts are constants
+    tokenX = addresses.tokenX;
+    tokenY = addresses.tokenY;
+
+    /**
+     * state which keeps track of total lqXY supply -- this is needed to calculate what to return when redeeming liquidity
+     *
+     * total supply is zero initially; it increases when supplying liquidity and decreases when redeeming it
+     */
+    @state(UInt64) totalSupply = State<UInt64>();
+
+    /**
+     * Mint liquidity tokens in exchange for X and Y tokens
+     * @param user caller address
+     * @param dx input amount of X tokens
+     * @param dy input amount of Y tokens
+     * @return output amount of lqXY tokens
+     *
+     * This function fails if the X and Y token amounts don't match the current X/Y ratio in the pool.
+     * This can also be used if the pool is empty. In that case, there is no check on X/Y;
+     * instead, the input X and Y amounts determine the initial ratio.
+     *
+     * The transaction needs to be signed by the user's private key.
+     */
+    @method supplyLiquidityBase(
+      user: PublicKey,
+      dx: UInt64,
+      dy: UInt64
+    ): UInt64 {
+      let tokenX = new TokenContract(this.tokenX);
+      let tokenY = new TokenContract(this.tokenY);
+
+      // get balances of X and Y token
+      // TODO: this creates extra account updates. we need to reuse these by passing them to or returning them from transfer()
+      // but for that, we need the @method argument generalization
+      let dexXBalance = tokenX.getBalance(this.address);
+      let dexYBalance = tokenY.getBalance(this.address);
+
+      // // assert dy == [dx * y/x], or x == 0
+      let isXZero = dexXBalance.equals(UInt64.zero);
+      let xSafe = Circuit.if(isXZero, UInt64.one, dexXBalance);
+
+      // FIXME
+      // Error: Constraint unsatisfied (unreduced): Equal 0 1
```

Comment:
Should we remove these comments now that the line below has been enabled? I'm also curious to know what was causing this error if you know the cause?

Reply:
I didn't test the line again with proving enabled - just wanted it enabled for fuller testing in the non proving case. All of these tests only work with not proving by now. I have an issue open just for investigating this line: https://github.com/o1-labs/snarkyjs/issues/499